### PR TITLE
Repair struct empty pointer filed for Java Integer Bool String...

### PR DESCRIPTION
**Pull request by Carlos-Yuan**
Repair struct empty pointer filed for Java Integer Bool String...

### CODE REVIEW

**Comment by wongoo** [comment]
@Carlos-Yuan does this pr based on https://github.com/dubbogo/hessian2/pull/109 ?  do you know @mcdull6226?

**Comment by wongoo** [comment]
@AlexStocks @fangyincheng @aliiohs @u0x01  pls review this pr.

**Comment by AlexStocks** [comment]
> @Carlos-Yuan does this pr based on #109 ? do you know @mcdull6226?

so amazing.

**Comment by Carlos-Yuan** [comment]
my other account

**Comment by Carlos-Yuan** [comment]
> @Carlos-Yuan does this pr based on #109 ? do you know @mcdull6226?
my other account

**Comment by wongoo** [comment]
@Carlos-Yuan https://github.com/dubbogo/hessian2/pull/109 can be closed if this pr already contains all that in it.

File: double.go [hunk 1]
```diff
@@ -82,6 +82,8 @@ func (d *Decoder) decDouble(flag int32) (interface{}, error) {
 		tag, _ = d.readByte()
 	}
 	switch tag {
+	case BC_NULL:
+		return nil, nil
```
Comment (fangyincheng) [inline double.go]: image

image

Are they different?

Comment (Carlos-Yuan) [reply double.go]: int.go 和 list.go 需要判断java返回的属性是不是空，然后在object.go里面对返回值进行判断，并赋值

File: codec.go [hunk 2]
```diff
@@ -383,39 +383,48 @@ func SetSlice(dest reflect.Value, objects interface{}) error {
 	if objects == nil {
 		return nil
 	}
+	if dest.Kind() == reflect.Ptr && dest.CanSet() {
+		destTyp := UnpackPtrType(dest.Type())
+		elemKind := destTyp.Elem().Kind()
+		if elemKind == reflect.Uint8 {
```
Comment (wongoo) [inline codec.go]: no handle for other situations

File: codec.go [hunk 3]
```diff
@@ -383,39 +383,48 @@ func SetSlice(dest reflect.Value, objects interface{}) error {
...
+			// for binary
+			bytes := objects.([]uint8)
+			dest.Set(reflect.ValueOf(&bytes))
+			return nil
+		}
+	} else {
```
Comment (wongoo) [inline codec.go]: not use if/else, return instead.

File: int.go [hunk 4]
```diff
@@ -98,6 +98,11 @@ func (d *Decoder) decInt32(flag int32) (int32, error) {
 		var i32 int32
 		err = binary.Read(d.reader, binary.BigEndian, &i32)
 		return i32, perrors.WithStack(err)
+	case tag == BC_OBJECT_DEF:
```
Comment (wongoo) [inline int.go]: the default case is already return error, is this needed?

File: list.go [hunk 5]
```diff
@@ -276,6 +276,10 @@ func (d *Decoder) decList(flag int32) (interface{}, error) {
 		return d.readTypedList(tag)
 	case untypedListTag(tag):
 		return d.readUntypedList(tag)
+	case tag == BC_BINARY_CHUNK:
```
Comment (wongoo) [inline list.go]: why binary tags exists in list?

File: long.go [hunk 6]
```diff
@@ -74,7 +74,7 @@ func (d *Decoder) decInt64(flag int32) (int64, error) {
 
 	switch {
 	case tag == BC_NULL:
-		return int64(0), nil
+		return int64(0), perrors.Errorf("Num is null")
```
Comment (wongoo) [inline long.go]: why return error for null?

File: object.go [hunk 7]
```diff
@@ -336,10 +336,22 @@ func (d *Decoder) decInstance(typ reflect.Type, cls classInfo) (interface{}, err
 			if err != nil {
 				return nil, perrors.Wrapf(err, "decInstance->ReadString: %s", fieldName)
 			}
-			fldRawValue.SetString(str)
+			if str != "null" {
```
Comment (wongoo) [inline object.go]: `null` can be the exact string data, don't agree with this logic

File: object.go [hunk 8]
```diff
@@ -350,19 +362,70 @@ func (d *Decoder) decInstance(typ reflect.Type, cls classInfo) (interface{}, err
 					}
 					enumValue, _ := s.(JavaEnum)
 					num = int32(enumValue)
+				} else if err.Error() == "Num is null" {
```
Comment (wongoo) [inline object.go]: not check the error message, define an error type instead.

File: object.go [hunk 9]
```diff
@@ -350,19 +362,70 @@ func (d *Decoder) decInstance(typ reflect.Type, cls classInfo) (interface{}, err
...
+					isNull = true
 				} else {
 					return nil, perrors.Wrapf(err, "decInstance->decInt32, field name:%s", fieldName)
 				}
 			}
-			fldRawValue.SetInt(int64(num))
+			if !isNull {
+				if fldRawValue.Kind() == reflect.Ptr && fldRawValue.CanSet() {
```
Comment (wongoo) [inline object.go]: it can be extracted to a new method `setIntValue`

File: object.go [hunk 10]
```diff
@@ -350,19 +362,70 @@ func (d *Decoder) decInstance(typ reflect.Type, cls classInfo) (interface{}, err
...
 			if err != nil {
-				return nil, perrors.Wrapf(err, "decInstance->decInt32, field name:%s", fieldName)
+				if err.Error() == "Num is null" {
+					isNull = true
+				} else {
+					return nil, perrors.Wrapf(err, "decInstance->decInt32, field name:%s", fieldName)
+				}
+			}
+			if !isNull {
+				if fldRawValue.Kind() == reflect.Ptr && fldRawValue.CanSet() {
```
Comment (wongoo) [inline object.go]: duplicated code

File: object.go [hunk 11]
```diff
@@ -372,31 +435,94 @@ func (d *Decoder) decInstance(typ reflect.Type, cls classInfo) (interface{}, err
 					}
 					enumValue, _ := s.(JavaEnum)
 					num = int64(enumValue)
+				} else if err.Error() == "Num is null" {
+					isNull = true
 				} else {
 					return nil, perrors.Wrapf(err, "decInstance->decInt64 field name:%s", fieldName)
 				}
 			}
+			if !isNull {
```
Comment (wongoo) [inline object.go]: duplicated code

**Comment by wongoo** [comment]
@Carlos-Yuan it's a big PR, it's hard to merge, if it's possible I suggest to split it to three:
- fix nil pointer for java Bool
- fix nil pointer for java Integer
- fix nil pointer for java String

File: double.go (hunk 1 above)
Comment (fangyincheng) [reply double.go]: ok

**Pull request closed without merging**